examples/marquee.lisp

```diff
@@ -0,0 +1,88 @@
+(defpackage #:charms-marquee
+  (:use #:cl)
+  (:export #:main))
+(in-package :charms-marquee)
+
+(defclass marquee ()
+  ((position :initarg :position :accessor marquee-position)
+   (size :initarg :size :accessor marquee-size)
+   (text :initarg :text :accessor marquee-text)
+   (viewport :initarg :viewport :reader marquee-viewport)
+   (padding-char :initarg :padding-char :initform #\Space :accessor marquee-padding-char)
+   (buffer :initform "" :accessor marquee-buffer))
+  (:documentation "A marquee is like a scrolling text thing. A string of characters moves through
+a viewport. The viewport constrains the text so you only see a little bit at a time."))
+
+;; (defmethod marquee-viewport :before ((marquee marquee))
+;;   "Before getting the MARQUEE viewport (the text that is shown at a given time), the buffer
+;; needs to be padded in case the text doesn't fill the MARQUEE size"
+;;  ( )
+
+(defmethod marquee-viewport ((marquee marquee))
+  "Evaluates to the text currently shown in the viewport for the given MARQUEE."
+  (marquee-pad-buffer marquee)
+  (subseq (slot-value marquee 'buffer) 0 (slot-value marquee 'size)))
+
+(defmethod marquee-pad-buffer ((marquee marquee))
+  "When the size of the MARQUEE is greater than the supplied text, the buffer needs to be padded
+with spaces."
+  (with-accessors ((text marquee-text) (size marquee-size) (buffer marquee-buffer) (pad-char marquee-padding-char))
+      marquee
+    ;; The buffer size can never be less than the size of the text. Otherwise, the text won't show.
+    (when (<= (length buffer) (length text))
+      (setf buffer text))
+    ;; This cond block pads the BUFFER. First condition is when the SIZE of the marquee is great than the TEXT. In this case,
+    ;; extra spaces are added to the
+    (cond ((> size (length buffer))
+           (setf buffer (concatenate 'string text (make-string (- size (length text)) :initial-element pad-char))))
+          ((and (< size (length buffer))
+                (> size (length text)))
```

Comment:
this is optional, but you can simplify the AND in this way:

```
(< (length text) size (length buffer))
```

This is more natural to a mathematically inclined person (like me) who reads $a < x < b$ inequalities all the time. But it may seem obscure otherwise. I'll let you judge which is clearer.